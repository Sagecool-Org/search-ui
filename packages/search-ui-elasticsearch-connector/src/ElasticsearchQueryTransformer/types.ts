import type {
  QueryDslQueryContainer,
  SearchRequest,
  SearchHighlight
} from "@elastic/elasticsearch/lib/api/types";

export type Filter = Pick<QueryDslQueryContainer, "bool">;
export type Sort = SearchRequest["sort"];

export type FilterValue = Pick<QueryDslQueryContainer, "term">;
export type FilterValueRange = Pick<QueryDslQueryContainer, "range">;

export type RequestBody = SearchRequest;

export type Query = Pick<QueryDslQueryContainer, "bool">;

export type HighlightFields = Pick<SearchHighlight, "fields">;
export type Aggregations = Record<string, any>;
