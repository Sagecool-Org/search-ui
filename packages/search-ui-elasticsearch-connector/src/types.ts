import type { estypes } from "@elastic/elasticsearch";
import { QueryConfig, RequestState } from "@elastic/search-ui";

export type SearchRequest = estypes.SearchRequest;
export type SearchSourceFilter = estypes.SearchSourceFilter;
export type SearchHit = estypes.SearchHit;

export type PostProcessRequestBodyFn = (
  requestBody: SearchRequest,
  requestState: RequestState,
  queryConfig: QueryConfig
) => SearchRequest;

export interface CloudHost {
  id: string;
}

export type ConnectionOptions = {
  host?: string;
  cloud?: CloudHost;
  index: string;
  apiKey?: string;
  connectionOptions?: {
    headers?: Record<string, string>;
  };
};
