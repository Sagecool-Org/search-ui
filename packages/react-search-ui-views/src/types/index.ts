import type { FacetValue, FieldValue, FilterType } from "@elastic/search-ui";

export interface BaseContainerProps {
  children?: React.ReactNode;
  className?: string;
}

export type { ErrorBoundaryViewProps } from "../ErrorBoundary";

export type { SearchBoxAutocompleteViewProps } from "../Autocomplete";

export type {
  PagingViewProps,
  PagingContainerContext,
  PagingContainerProps
} from "../Paging";

export type {
  PagingInfoContainerProps,
  PagingInfoContainerContext,
  PagingInfoViewProps
} from "../PagingInfo";

export type { ResultsContainerProps, ResultsViewProps } from "../Results";

export type { ResultContainerProps, ResultViewProps } from "../Result";

export type {
  ResultsPerPageContainerContext,
  ResultsPerPageContainerProps,
  ResultsPerPageViewProps
} from "../ResultsPerPage";

export type {
  SearchBoxContainerContext,
  SearchBoxContainerProps,
  SearchBoxViewProps
} from "../SearchBox";

export type { InputViewProps } from "../SearchInput";

export type { SortingContainerProps, SortingViewProps } from "../Sorting";

export type FacetViewProps = {
  className?: string;
  label: string;
  onMoreClick: () => void;
  onRemove: (value: FieldValue) => void;
  onChange: (value: FieldValue) => void;
  onSelect: (value: FieldValue) => void;
  options: FacetValue[];
  showMore: boolean;
  values: any;
  showSearch: boolean;
  onSearch: (value: string) => void;
  searchPlaceholder: string;
};

export type FacetContainerProps = BaseContainerProps & {
  filterType?: FilterType;
  show?: number;
  view?: React.ComponentType<FacetViewProps>;
  isFilterable?: boolean;
  field: string;
  label: string;
  persistent?: boolean;
};

// From SO https://stackoverflow.com/a/59071783
// TS Utility to rename keys in a type
// Usage type x = Rename<{ renameMe: string }, { "renameMe": "newName" }>
// x === { newName: string }
export type Rename<
  T,
  R extends {
    [K in keyof R]: K extends keyof T ? PropertyKey : "Error: key not in T";
  }
> = { [P in keyof T as P extends keyof R ? R[P] : P]: T[P] };
