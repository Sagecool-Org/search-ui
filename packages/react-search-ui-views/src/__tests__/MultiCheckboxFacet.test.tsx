import React from "react";
import { render } from "@testing-library/react";
import MultiCheckboxFacet from "../MultiCheckboxFacet";
import { FacetViewProps } from "../types";

const params: FacetViewProps = {
  label: "A Facet",
  onMoreClick: jest.fn(),
  onRemove: jest.fn(),
  onSelect: jest.fn(),
  onSearch: jest.fn(),
  onChange: jest.fn(),
  searchPlaceholder: "Search",
  showSearch: false,
  values: [],
  options: [
    {
      value: "fieldValue1",
      count: 10,
      selected: false
    },
    {
      value: "fieldValue2",
      count: 5,
      selected: true
    }
  ],
  showMore: true
};

const rangeOptions = [
  {
    count: 1,
    value: {
      from: 1,
      to: 10,
      name: "The first option"
    }
  },
  {
    count: 11,
    value: {
      from: 11,
      to: 20,
      name: "The second option"
    }
  }
];

it("renders", () => {
  const { container } = render(<MultiCheckboxFacet {...params} />);
  expect(container).toMatchSnapshot();
});

it("renders falsey values correctly", () => {
  const { container } = render(
    <MultiCheckboxFacet
      {...params}
      options={[
        {
          value: 0,
          count: 10,
          selected: false
        },
        {
          value: false,
          count: 20,
          selected: false
        },
        {
          value: "",
          count: 30,
          selected: false
        }
      ]}
    />
  );
  expect(container).toMatchSnapshot();
});

it("renders range filters", () => {
  const { container } = render(
    <MultiCheckboxFacet {...params} options={rangeOptions} />
  );
  expect(container).toMatchSnapshot();
});

it("will render 'more' button if more param is true", () => {
  const { container } = render(
    <MultiCheckboxFacet
      {...{
        ...params,
        showMore: true
      }}
    />
  );
  expect(container.querySelector(".sui-facet-view-more")).toBeInTheDocument();
});

it("will render a no results message is no options are available", () => {
  const { container } = render(
    <MultiCheckboxFacet
      {...{
        ...params,
        options: []
      }}
    />
  );
  expect(
    container.querySelector(".sui-multi-checkbox-facet")
  ).toHaveTextContent("No matching options");
});

it("won't render 'more' button if more param is false", () => {
  const { container } = render(
    <MultiCheckboxFacet
      {...{
        ...params,
        showMore: false
      }}
    />
  );
  expect(
    container.querySelector(".sui-multi-checkbox-facet__view-more")
  ).not.toBeInTheDocument();
});

describe("determining selected option", () => {
  it("will correctly determine which of the options is selected", () => {
    const { container } = render(<MultiCheckboxFacet {...params} />);
    const checkboxes = container.querySelectorAll("input[type='checkbox']");
    expect(checkboxes[0]).not.toBeChecked();
    expect(checkboxes[1]).toBeChecked();
  });
});

it("renders with className prop applied", () => {
  const customClassName = "test-class";
  const { container } = render(
    <MultiCheckboxFacet className={customClassName} {...params} />
  );
  expect(container.firstChild).toHaveClass("sui-facet", "test-class");
});

it("will render search input if `showSearch` param is true", () => {
  const { container } = render(
    <MultiCheckboxFacet
      {...{
        ...params,
        showSearch: true
      }}
    />
  );
  expect(container.querySelector(".sui-facet-search")).toBeInTheDocument();
});

it("won't render search input if `showSearch` param is false", () => {
  const { container } = render(
    <MultiCheckboxFacet
      {...{
        ...params,
        showSearch: false
      }}
    />
  );
  expect(container.querySelector(".sui-facet-search")).not.toBeInTheDocument();
});

it("should use the `searchPlaceholder` param as a search input placeholder", () => {
  const searchPlaceholder = "bingo";
  const { container } = render(
    <MultiCheckboxFacet
      {...{
        ...params,
        showSearch: true,
        searchPlaceholder
      }}
    />
  );
  expect(
    container.querySelector(".sui-facet-search__text-input")
  ).toHaveAttribute("placeholder", searchPlaceholder);
});
