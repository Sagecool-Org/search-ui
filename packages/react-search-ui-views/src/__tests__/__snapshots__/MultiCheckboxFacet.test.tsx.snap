// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <fieldset
    class="sui-facet"
  >
    <legend
      class="sui-facet__title"
    >
      A Facet
    </legend>
    <div
      class="sui-multi-checkbox-facet"
    >
      <label
        class="sui-multi-checkbox-facet__option-label"
        for="example_facet_A FacetfieldValue1"
      >
        <div
          class="sui-multi-checkbox-facet__option-input-wrapper"
        >
          <input
            class="sui-multi-checkbox-facet__checkbox"
            data-transaction-name="facet - A Facet"
            id="example_facet_A FacetfieldValue1"
            type="checkbox"
          />
          <span
            class="sui-multi-checkbox-facet__input-text"
          >
            fieldValue1
          </span>
        </div>
        <span
          class="sui-multi-checkbox-facet__option-count"
        >
          10
        </span>
      </label>
      <label
        class="sui-multi-checkbox-facet__option-label"
        for="example_facet_A FacetfieldValue2"
      >
        <div
          class="sui-multi-checkbox-facet__option-input-wrapper"
        >
          <input
            checked=""
            class="sui-multi-checkbox-facet__checkbox"
            data-transaction-name="facet - A Facet"
            id="example_facet_A FacetfieldValue2"
            type="checkbox"
          />
          <span
            class="sui-multi-checkbox-facet__input-text"
          >
            fieldValue2
          </span>
        </div>
        <span
          class="sui-multi-checkbox-facet__option-count"
        >
          5
        </span>
      </label>
    </div>
    <button
      aria-label="Show more options"
      class="sui-facet-view-more"
      type="button"
    >
      + More
    </button>
  </fieldset>
</div>
`;

exports[`renders falsey values correctly 1`] = `
<div>
  <fieldset
    class="sui-facet"
  >
    <legend
      class="sui-facet__title"
    >
      A Facet
    </legend>
    <div
      class="sui-multi-checkbox-facet"
    >
      <label
        class="sui-multi-checkbox-facet__option-label"
        for="example_facet_A Facet0"
      >
        <div
          class="sui-multi-checkbox-facet__option-input-wrapper"
        >
          <input
            class="sui-multi-checkbox-facet__checkbox"
            data-transaction-name="facet - A Facet"
            id="example_facet_A Facet0"
            type="checkbox"
          />
          <span
            class="sui-multi-checkbox-facet__input-text"
          >
            0
          </span>
        </div>
        <span
          class="sui-multi-checkbox-facet__option-count"
        >
          10
        </span>
      </label>
      <label
        class="sui-multi-checkbox-facet__option-label"
        for="example_facet_A Facetfalse"
      >
        <div
          class="sui-multi-checkbox-facet__option-input-wrapper"
        >
          <input
            class="sui-multi-checkbox-facet__checkbox"
            data-transaction-name="facet - A Facet"
            id="example_facet_A Facetfalse"
            type="checkbox"
          />
          <span
            class="sui-multi-checkbox-facet__input-text"
          >
            false
          </span>
        </div>
        <span
          class="sui-multi-checkbox-facet__option-count"
        >
          20
        </span>
      </label>
      <label
        class="sui-multi-checkbox-facet__option-label"
        for="example_facet_A Facet"
      >
        <div
          class="sui-multi-checkbox-facet__option-input-wrapper"
        >
          <input
            class="sui-multi-checkbox-facet__checkbox"
            data-transaction-name="facet - A Facet"
            id="example_facet_A Facet"
            type="checkbox"
          />
          <span
            class="sui-multi-checkbox-facet__input-text"
          />
        </div>
        <span
          class="sui-multi-checkbox-facet__option-count"
        >
          30
        </span>
      </label>
    </div>
    <button
      aria-label="Show more options"
      class="sui-facet-view-more"
      type="button"
    >
      + More
    </button>
  </fieldset>
</div>
`;

exports[`renders range filters 1`] = `
<div>
  <fieldset
    class="sui-facet"
  >
    <legend
      class="sui-facet__title"
    >
      A Facet
    </legend>
    <div
      class="sui-multi-checkbox-facet"
    >
      <label
        class="sui-multi-checkbox-facet__option-label"
        for="example_facet_A FacetThe first option"
      >
        <div
          class="sui-multi-checkbox-facet__option-input-wrapper"
        >
          <input
            class="sui-multi-checkbox-facet__checkbox"
            data-transaction-name="facet - A Facet"
            id="example_facet_A FacetThe first option"
            type="checkbox"
          />
          <span
            class="sui-multi-checkbox-facet__input-text"
          >
            The first option
          </span>
        </div>
        <span
          class="sui-multi-checkbox-facet__option-count"
        >
          1
        </span>
      </label>
      <label
        class="sui-multi-checkbox-facet__option-label"
        for="example_facet_A FacetThe second option"
      >
        <div
          class="sui-multi-checkbox-facet__option-input-wrapper"
        >
          <input
            class="sui-multi-checkbox-facet__checkbox"
            data-transaction-name="facet - A Facet"
            id="example_facet_A FacetThe second option"
            type="checkbox"
          />
          <span
            class="sui-multi-checkbox-facet__input-text"
          >
            The second option
          </span>
        </div>
        <span
          class="sui-multi-checkbox-facet__option-count"
        >
          11
        </span>
      </label>
    </div>
    <button
      aria-label="Show more options"
      class="sui-facet-view-more"
      type="button"
    >
      + More
    </button>
  </fieldset>
</div>
`;
